### PR TITLE
add CodeIssue check type

### PR DESCRIPTION
**Pull request by davidbloss**
Resolves [#504](https://github.com/OpsLevel/team-platform/issues/504)

Added CodeIssue check and update check tests to be more robust

### Checklist

- [ ] I have run this code, and it appears to resolve the stated issue.
- [ ] This PR does not reduce total test coverage
- [ ] This PR has no user interface changes or has already received approval from product management to change the interface.
- [ ] Does this change require a Terraform schema change?
  - If so what is the ticket or PR #
- [ ] Make a [changie](https://github.com/OpsLevel/opslevel-go/blob/main/CONTRIBUTING.md#changie-change-log-generation) entry that explains the customer facing outcome of this change

### CODE REVIEW

File: check_code_issue.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package opslevel
+
+type CodeIssueCheckFragment struct {
+	Campaign string `graphql:"campaign"` // The campaign the check belongs to.
```
Comment (rocktavious) [inline check_code_issue.go:4]: This does not feel correct - why was this added?  (This feels like its a field on the base check type)

File: input.go [hunk 2]
```diff
@@ -101,6 +101,43 @@ type CheckAlertSourceUsageUpdateInput struct {
...
+	LevelId        ID                            `json:"levelId" yaml:"levelId" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`                       // The id of the level the check belongs to. (Required.)
+	OwnerId        *ID                           `json:"ownerId,omitempty" yaml:"ownerId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`   // The id of the team that owns the check. (Optional.)
+	FilterId       *ID                           `json:"filterId,omitempty" yaml:"filterId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"` // The id of the filter of the check. (Optional.)
+	Notes          *string                       `json:"notes,omitempty" yaml:"notes,omitempty" example:"example_notes"`                         // Additional information about the check. (Optional.)
+	Constraint     CheckCodeIssueConstraintEnum  `json:"constraint" yaml:"constraint" example:"NEW_ENUM_SET_DEFAULT"`                            // The type of constraint used in evaluation the code issues check. (Required.)
```
Comment (rocktavious) [inline input.go:114]: The `example` tag is wrong?

File: input.go [hunk 3]
```diff
@@ -595,6 +632,12 @@ type ChecksCopyToCampaignInput struct {
 	CheckIds   []string `json:"checkIds" yaml:"checkIds" example:"['Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk', 'Z2lkOi8vc2VydmljZS85ODc2NTQzMjE']"` // The IDs of the existing rubric checks to be copied. (Required.)
 }
 
+// CodeIssueResolutionTimeInput represents the allowed threshold for how long an issue has been detected before the check starts failing.
+type CodeIssueResolutionTimeInput struct {
+	Unit  CodeIssueResolutionTimeUnitEnum `json:"unit" yaml:"unit" example:"NEW_ENUM_SET_DEFAULT"` // . (Required.)
+	Value int                             `json:"value" yaml:"value" example:"3"`                  // . (Required.)
```
Comment (rocktavious) [inline input.go:638]: feels like we are missing docstrings on this - is that an API problem?

File: check_code_issue.go (hunk 1 above)
Comment (rocktavious) [reply check_code_issue.go:4]: Also where is `Constraint` and `IssueName` and `ResolutionTime` and `MaxAllowed` etc etc.

These are surely in the response payload.

File: check_code_issue.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	Url      string `graphql:"url"`      // The url to the check.
```
Comment (rocktavious) [inline check_code_issue.go:5]: Same with this field - this is likely on the base check type not specific to the CodeIssuesCheck type.

**Review by rocktavious — CHANGES_REQUESTED** [review]
Lots of issues with this one.  Let me know if you want to sit down and chat through it.

File: check_code_issue.go (hunk 1 above)
Comment (davidbloss) [reply check_code_issue.go:4]: Oddly enough they are not. I'll ask around and see if these were forgotten in the API

File: check_code_issue.go (hunk 4 above)
Comment (davidbloss) [reply check_code_issue.go:5]: You're right on. `Campaign` and `Url` are both part of the base check type

File: input.go (hunk 2 above)
Comment (davidbloss) [reply input.go:114]: Yes and good call!

File: input.go [hunk 5]
```diff
@@ -595,6 +632,12 @@ type ChecksCopyToCampaignInput struct {
 	CheckIds   []string `json:"checkIds" yaml:"checkIds" example:"['Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk', 'Z2lkOi8vc2VydmljZS85ODc2NTQzMjE']"` // The IDs of the existing rubric checks to be copied. (Required.)
 }
 
+// CodeIssueResolutionTimeInput represents the allowed threshold for how long an issue has been detected before the check starts failing.
+type CodeIssueResolutionTimeInput struct {
+	Unit  CodeIssueResolutionTimeUnitEnum `json:"unit" yaml:"unit" example:"NEW_ENUM_SET_DEFAULT"` // . (Required.)
```
[the remark this answers is not in the document]
Comment (davidbloss) [reply input.go:637]: Yes wrong here too

File: check_test.go [hunk 6]
```diff
@@ -258,6 +278,73 @@ func getCheckTestCases() map[string]TmpCheckTestCase {
...
+			expectedCheck: CheckWithExtras(map[string]any{
+				"constraint":     "exact",
+				"issueName":      "test-issue",
+				"issueType":      []string{"bug", "error"},
+				"maxAllowed":     3,
+				"resolutionTime": map[string]any{"unit": "day", "value": 1},
+				"severity":       []string{"sev1", "sev2"},
+			}),
+		},
+		"UpdateCodeIssue": {
```
Comment (rocktavious) [inline check_test.go:318]: I think we'll want to throw a few more "update" test cases in here to ensure it generates the right input for a few of these optional fields.  `maxAllowed`, `resolutionTime` etc etc

**Review by rocktavious — APPROVED** [review]
If the extra tests are too tricky to write correctly because of API validation - then this is 👍

File: check_code_issue.go [hunk 7]
```diff
@@ -2,10 +2,10 @@ package opslevel
 
 type CodeIssueCheckFragment struct {
 	Constraint     CheckCodeIssueConstraintEnum `graphql:"constraint"`     // The type of constraint used in evaluation the code issues check.
-	IssueName      string                       `graphql:"issueName"`      // The issue name used for code issue lookup.
+	IssueName      *string                      `graphql:"issueName"`      // The issue name used for code issue lookup.
 	IssueType      []string                     `graphql:"issueType"`      // The type of code issue to consider.
-	MaxAllowed     int                          `graphql:"maxAllowed"`     // The threshold count of code issues beyond which the check starts failing.
-	ResolutionTime CodeIssueResolutionTime      `graphql:"resolutionTime"` // The resolution time recommended by the reporting source of the code issue.
+	MaxAllowed     *int                         `graphql:"maxAllowed"`     // The threshold count of code issues beyond which the check starts failing.
+	ResolutionTime *CodeIssueResolutionTime     `graphql:"resolutionTime"` // The resolution time recommended by the reporting source of the code issue.
```
Comment (rocktavious) [inline check_code_issue.go:8]: What caused us to change these to pointers?

Comment (davidbloss) [reply check_code_issue.go]: Classic whoopsie daisy

**Review by rocktavious — APPROVED** [review]
Heafty but good work!!
